File: advanced_auth/connected_apps/b2b-integrated/sprintplanner-python-fastmcp-backend/mcp_server.py
"""
Standalone MCP Server for the Ticket Board application
"""

from dotenv import load_dotenv
from fastmcp import FastMCP
from fastmcp.server.auth import BearerAuthProvider
from fastmcp.server.dependencies import get_access_token
from typing import List, Dict, Any, Optional
import crud
import schemas
import os


load_dotenv(".env.local")

auth = BearerAuthProvider(
    jwks_uri=f"{os.getenv('STYTCH_DOMAIN')}/.well-known/jwks.json",
    issuer=os.getenv("STYTCH_DOMAIN"),
    algorithm="RS256",
    audience=os.getenv("STYTCH_PROJECT_ID")
)

# Create FastMCP instance
mcp = FastMCP("ticket-board-mcp", auth=auth)

def get_organization_id_from_context() -> str:
    """Extract organization ID from the current session context"""
    # This will be populated by the auth system when a request comes in
    # For now, we'll get it from the access token
    token = get_access_token()
    if not token:
        raise ValueError("No access token found in context")

    return token.claims.get("https://stytch.com/organization", {}).get("organization_id")

@mcp.tool()
async def list_tickets() -> List[Dict[str, Any]]:
    """List all tickets for the authenticated organization"""
    organization_id = get_organization_id_from_context()
    tickets = crud.get_tickets(organization_id)
    return [
        {
            "id": ticket.id,
            "title": ticket.title,
            "assignee": ticket.assignee,
            "status": ticket.status,
            "description": ticket.description,
            "created_at": ticket.created_at.isoformat(),
            "updated_at": ticket.updated_at.isoformat()
        }
        for ticket in tickets
    ]

@mcp.tool()
async def get_ticket(ticket_id: str) -> Optional[Dict[str, Any]]:
    """Get a specific ticket by ID for the authenticated organization"""
    organization_id = get_organization_id_from_context()
    ticket = crud.get_ticket(ticket_id, organization_id)
    if ticket:
        return {
            "id": ticket.id,
            "title": ticket.title,
            "assignee": ticket.assignee,
            "status": ticket.status,
            "description": ticket.description,
            "created_at": ticket.created_at.isoformat(),
            "updated_at": ticket.updated_at.isoformat()
        }
    return None

@mcp.tool()
async def create_ticket(
    title: str, 
    assignee: str, 
    description: Optional[str] = None
) -> Dict[str, Any]:
    """Create a new ticket for the authenticated organization"""
    organization_id = get_organization_id_from_context()
    
    # Ensure organization exists
    crud.get_or_create_organization(organization_id)
    
    # Create ticket data
    ticket_data = schemas.TicketCreate(
        title=title,
        assignee=assignee,
        description=description
    )
    
    # Create the ticket
    ticket = crud.create_ticket(ticket_data, organization_id)
    
    return {
        "id": ticket.id,
        "title": ticket.title,
        "assignee": ticket.assignee,
        "status": ticket.status,
        "description": ticket.description,
        "created_at": ticket.created_at.isoformat(),
        "updated_at": ticket.updated_at.isoformat()
    }

@mcp.tool()
async def update_ticket_status(
    ticket_id: str, 
    status: str
) -> Optional[Dict[str, Any]]:
    """Update the status of a ticket"""
    organization_id = get_organization_id_from_context()
    ticket = crud.update_ticket_status(ticket_id, status, organization_id)
    if ticket:
        return {
            "id": ticket.id,
            "title": ticket.title,
            "assignee": ticket.assignee,
            "status": ticket.status,
            "description": ticket.description,
            "created_at": ticket.created_at.isoformat(),
            "updated_at": ticket.updated_at.isoformat()
        }
    return None

@mcp.tool()
async def delete_ticket(ticket_id: str) -> bool:
    """Delete a ticket from the authenticated organization"""
    organization_id = get_organization_id_from_context()
    return crud.delete_ticket(ticket_id, organization_id)



@mcp.tool()
async def get_organization(organization_id: str) -> Optional[Dict[str, Any]]:
    """Get a specific organization by ID"""
    org = crud.get_organization(organization_id)
    info = org.name if org else None
    return info

@mcp.tool()
async def search_tickets(
    status: Optional[str] = None,
    assignee: Optional[str] = None,
    title_contains: Optional[str] = None
) -> List[Dict[str, Any]]:
    """Search tickets with filters for the authenticated organization"""
    organization_id = get_organization_id_from_context()
    tickets = crud.search_tickets(
        organization_id,
        status=status,
        assignee=assignee,
        title_contains=title_contains,
    )
    
    return [
        {
            "id": ticket.id,
            "title": ticket.title,
            "assignee": ticket.assignee,
            "status": ticket.status,
            "description": ticket.description,
            "created_at": ticket.created_at.isoformat(),
            "updated_at": ticket.updated_at.isoformat()
        }
        for ticket in tickets
    ]

@mcp.tool()
async def get_ticket_statistics() -> Dict[str, Any]:
    """Get statistics about tickets for the authenticated organization"""
    organization_id = get_organization_id_from_context()
    tickets = crud.get_tickets(organization_id)
    
    total_tickets = len(tickets)
    status_counts = {}
    assignee_counts = {}
    
    for ticket in tickets:
        # Count by status
        status_counts[ticket.status] = status_counts.get(ticket.status, 0) + 1
        
        # Count by assignee
        assignee_counts[ticket.assignee] = assignee_counts.get(ticket.assignee, 0) + 1
    
    return {
        "total_tickets": total_tickets,
        "status_distribution": status_counts,
        "assignee_distribution": assignee_counts,
        "organization_id": organization_id
    }


@mcp.resource("tickets://authenticated", mime_type="application/json")
async def tickets_resource() -> List[Dict[str, Any]]:
    organization_id = get_organization_id_from_context()
    tickets = crud.get_tickets(organization_id)
    return [
        {
            "id": t.id,
            "title": t.title,
            "assignee": t.assignee,
            "status": t.status,
            "description": t.description,
            "created_at": t.created_at.isoformat() if t.created_at else None,
            "updated_at": t.updated_at.isoformat() if t.updated_at else None,
        }
        for t in tickets
    ]

File: advanced_auth/connected_apps/b2b-integrated/sprintplanner-express-mcpsdk-backend/src/mcpServer.ts
import { McpServer, ResourceTemplate } from '@modelcontextprotocol/sdk/server/mcp.js';
import { z } from 'zod';
import { TicketService } from './TicketService.js';
import { Ticket, TicketCreate, TicketSearchParams } from './types.js';

function formatResponse(description: string, tickets: Ticket[]) {
  return {
    content: [
      {
        type: 'text' as const,
        text: `Success! ${description}\n\nTickets:\n${JSON.stringify(tickets, null, 2)}`,
      },
    ],
  };
}

function formatSingleTicketResponse(description: string, ticket: Ticket | null) {
  return {
    content: [
      {
        type: 'text' as const,
        text: `Success! ${description}\n\nTicket:\n${ticket ? JSON.stringify(ticket, null, 2) : 'null'}`,
      },
    ],
  };
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function formatStatisticsResponse(description: string, statistics: any) {
  return {
    content: [
      {
        type: 'text' as const,
        text: `Success! ${description}\n\nStatistics:\n${JSON.stringify(statistics, null, 2)}`,
      },
    ],
  };
}

export function createMcpServer(ticketService: TicketService): McpServer {
  const server = new McpServer({ name: 'Ticket Board Service', version: '1.0.0' });

  // Resource: Tickets
  server.resource(
    'Tickets',
    new ResourceTemplate('tickets://authenticated/{id}', {
      list: async () => {
        const tickets = ticketService.getTickets();
        return {
          resources: tickets.map((ticket) => ({
            name: ticket.title,
            uri: `tickets://authenticated/${ticket.id}`,
          })),
        };
      },
    }),
    async (uri, { id }) => {
      if (typeof id !== 'string') {
        throw new Error('Invalid ticket ID');
      }
      const ticket = ticketService.getTicket(id);
      return {
        contents: [
          {
            uri: uri.href,
            text: ticket
              ? `ID: ${ticket.id}\nTitle: ${ticket.title}\nAssignee: ${ticket.assignee}\nStatus: ${ticket.status}\nDescription: ${ticket.description || 'No description'}\nCreated: ${ticket.created_at.toISOString()}\nUpdated: ${ticket.updated_at.toISOString()}`
              : 'NOT FOUND',
          },
        ],
      };
    },
  );

  // Tool: List all tickets
  server.tool('list_tickets', 'List all tickets for the authenticated organization', {}, async () => {
    const tickets = ticketService.getTickets();
    return formatResponse('Listed all tickets', tickets);
  });

  // Tool: Get specific ticket
  server.tool(
    'get_ticket',
    'Get a specific ticket by ID',
    { ticket_id: z.string().describe('The ID of the ticket to retrieve') },
    async ({ ticket_id }) => {
      if (!ticket_id) {
        throw new Error('ticket_id is required');
      }
      const ticket = ticketService.getTicket(ticket_id);
      return formatSingleTicketResponse('Retrieved ticket', ticket);
    },
  );

  // Tool: Create ticket
  server.tool(
    'create_ticket',
    'Create a new ticket',
    {
      title: z.string().describe('The title of the ticket'),
      assignee: z.string().describe('The person assigned to the ticket'),
      description: z.string().optional().describe('Optional description of the ticket'),
    },
    async ({ title, assignee, description }) => {
      const ticketData: TicketCreate = {
        title,
        assignee,
        ...(description !== undefined ? { description } : {}),
      };
      const newTicket = ticketService.createTicket(ticketData);
      return formatSingleTicketResponse('Created new ticket', newTicket);
    },
  );

  // Tool: Update ticket status
  server.tool(
    'update_ticket_status',
    'Update the status of a ticket',
    {
      ticket_id: z.string().describe('The ID of the ticket to update'),
      status: z.enum(['backlog', 'in-progress', 'review', 'done']).describe('The new status for the ticket'),
    },
    async ({ ticket_id, status }) => {
      const updatedTicket = ticketService.updateTicketStatus(ticket_id, status);
      return formatSingleTicketResponse('Updated ticket status', updatedTicket);
    },
  );

  // Tool: Delete ticket
  server.tool(
    'delete_ticket',
    'Delete a ticket',
    { ticket_id: z.string().describe('The ID of the ticket to delete') },
    async ({ ticket_id }) => {
      const success = ticketService.deleteTicket(ticket_id);
      return {
        content: [
          {
            type: 'text' as const,
            text: success ? `Success! Deleted ticket ${ticket_id}` : `Failed! Ticket ${ticket_id} not found`,
          },
        ],
      };
    },
  );

  // Tool: Search tickets
  server.tool(
    'search_tickets',
    'Search tickets with filters',
    {
      status: z.string().optional().describe('Filter by status (backlog, in-progress, review, done)'),
      assignee: z.string().optional().describe('Filter by assignee name'),
      title_contains: z.string().optional().describe('Filter by title containing text'),
    },
    async ({ status, assignee, title_contains }) => {
      const searchParams: TicketSearchParams = {
        ...(status !== undefined ? { status } : {}),
        ...(assignee !== undefined ? { assignee } : {}),
        ...(title_contains !== undefined ? { title_contains } : {}),
      };
      const tickets = ticketService.searchTickets(searchParams);
      return formatResponse('Searched tickets', tickets);
    },
  );

  // Tool: Get ticket statistics
  server.tool('get_ticket_statistics', 'Get statistics about tickets for the organization', {}, async () => {
    const statistics = ticketService.getTicketStatistics();
    return formatStatisticsResponse('Retrieved ticket statistics', statistics);
  });

  // Tool: Get organization info
  server.tool('get_organization', 'Get information about the current organization', {}, async () => {
    const organization = ticketService.getOrganization();
    return {
      content: [
        {
          type: 'text' as const,
          text: organization ? `Organization: ${organization.name} (ID: ${organization.id})` : 'Organization not found',
        },
      ],
    };
  });

  return server;
}
